### PR TITLE
Add CNTK as keras backend

### PR DESCRIPTION
**Pull request by souptc**
This is the Beta version of CNTK as keras backend. Those changes are suppose to work with CNTK v2.0 GA. Because the CNTKv2.0 is not officially published, you could use the wheel below to have a first try:

http://cntk.ai/PythonWheel/ForKeras/cntk-2.0rc3-cp35-cp35m-linux_x86_64.whl
http://cntk.ai/PythonWheel/ForKeras/cntk-2.0rc3-cp27-cp27mu-linux_x86_64.whl

Most of the keras features are supported, except:

1. Performance optimization on CPU device.
CNTK performance optimization on cpu device is not finished, so if you want better performance, please try to run CNTK_Keras with GPU device. 
 
2. Gradient as symbolic ops.
This is not include in CNTK_Keras beta release, and we will support it in CNTK_Keras GA release. 
 
3. Stateful recurrent layer.
This is not include in CNTK_Keras beta release. 
 
4. Masking on recurrent layer.
This is not include in CNTK_Keras beta release, and we will support it in CNTK_Keras GA release. 
 
5. Padding with non-specified shape. 
Padding with non-speicfied shape is not supported now. To using cntk backend in keras with padding, please specified the concrete input shape. 
 
6. Convolution with dilation.
This is not include in CNTK_Keras beta release, and we will support it in CNTK_Keras GA release. 
 
7. Randomness op across batch axis. 
This is not include in CNTK_Keras beta release, and we will support it in CNTK_Keras GA release. 
 
8. Some Keras backend apis: reverse / top_k / ctc / map / foldl / foldr. 
These are not include in CNTK_Keras beta release, and we will support it in CNTK_Keras GA release.

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -8,6 +8,10 @@ All contributions by Google:
 Copyright (c) 2015, Google, Inc.
 All rights reserved.
 
+All contributions by Microsoft:
+Copyright (c) 2015, Google, Inc.
```
Comment (Dref360) [inline LICENSE]: Microsoft not Google I guess?

File: docs/templates/backend.md [hunk 2]
```diff
@@ -84,7 +85,7 @@ The code below instantiates an input placeholder. It's equivalent to `tf.placeho
 input = K.placeholder(shape=(2, 4, 5))
 # also works:
 input = K.placeholder(shape=(None, 4, 5))
-# also works:
+# also works:m,
```
Comment (Dref360) [inline docs/templates/backend.md]: Remove

File: examples/conv_lstm.py [hunk 3]
```diff
@@ -8,11 +8,18 @@
...
 
 # We create a layer which take as input movies of shape
 # (n_frames, width, height, channels) and returns a movie
 # of identical shape.
 
+
+if K.backend() == 'cntk':
+    raise RuntimeError('CNTK could not run this example, due to the missing support for padding '
+                       'with non-specified input shape. To run it with CNTK, please update the '
+                       'input_shape in top COnvLSTM2D layer from (None, 40, 40, 1) to (15, 40, 40, 1)')
```
Comment (Dref360) [inline examples/conv_lstm.py]: Conv

Comment (Dref360) [inline examples/conv_lstm.py]: Specify a shape and then check if the shape is okay. Otherwise, even if I set input_shape like yu say, the exception still throw.

input_shape=(15,40,40,1)
if K.backend()=='cntk' and input_shape[0] is None:
   raise...

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -3638,3 +3638,46 @@ def foldr(fn, elems, initializer=None, name=None):
         Same type and shape as initializer
     """
     return tf.foldr(fn, elems, initializer=initializer, name=name)
+
+
+def dropout_on_constant_mask(inputs, dim, dropout_value, training):
+    """Generate a drop out mask as the constant during recurrent.
```
Comment (Dref360) [inline keras/backend/tensorflow_backend.py]: Separate PR?

File: keras/backend/theano_backend.py [hunk 5]
```diff
@@ -2282,3 +2282,22 @@ def foldr(fn, elems, initializer=None, name=None):
     fn2 = lambda x, acc: fn(acc, x)
 
     return theano.foldr(fn2, elems, initializer, name=name)[0]
+
+
+def dropout_on_constant_mask(inputs, dim, dropout_value, training):
+    ones = ones_like(reshape(inputs[:, 0, 0], (-1, 1)))
```
Comment (Dref360) [inline keras/backend/theano_backend.py]: Separate PR?

File: README.md [hunk 6]
```diff
@@ -125,11 +125,21 @@ Keras uses the following dependencies:
...
+
+- CNTK
+    - [See installation instructions](https://docs.microsoft.com/en-us/cognitive-toolkit/setup-cntk-on-your-machine).
+
 *When using the Theano backend:*
 
 - Theano
     - [See installation instructions](http://deeplearning.net/software/theano/install.html#install).
 
+*When using the CNTK backend:*
```
Comment (n17s) [inline README.md]: duplicate

File: LICENSE (hunk 1 above)
Comment (souptc) [inline LICENSE]: Thanks! Fixed.

File: docs/templates/backend.md (hunk 2 above)
Comment (souptc) [inline docs/templates/backend.md]: Fixed.

File: examples/conv_lstm.py (hunk 3 above)
Comment (souptc) [inline examples/conv_lstm.py]: Because this is a beta version of cntk backend, parts of the examples are not supported. Those "if" codes were here to give user better message. After consider the impact, we revert the changes in examples, instead we provide better error message in cntk_backend, to avoid the complexity for keras management.

CNTK will fix the padding issue in GA version of keras backend.

File: keras/backend/tensorflow_backend.py (hunk 4 above)
Comment (souptc) [inline keras/backend/tensorflow_backend.py]: These two new methods introduced is because cntk's way to handle the drop out mask is a little different as the existing keras code, so we did a small refactoring to avoid add "if K.backend() == cntk" statement in keras recurrent layer. There is no new feature in tensorflow_backend and theano_backend.  Do we prefer to have the refactoring of recurrent layer with the two new method in a separate PR?

File: keras/backend/theano_backend.py (hunk 5 above)
Comment (souptc) [inline keras/backend/theano_backend.py]: same as the reply in tensorflow_backend.

File: README.md (hunk 6 above)
Comment (souptc) [inline README.md]: Fixed.

File: examples/conv_lstm.py (hunk 3 above)
Comment (souptc) [inline examples/conv_lstm.py]: For your concern of too many "if" into keras code. we already tried to wrapper most changes in cntk backend, to avoid add "if" to the keras layer code. The only exception is the local layer, because the way cntk support it is totally different... If keras has strong concern about that part, we can have more detail discussion (like try to support the keras way in the next CNTK Keras GA release). If you have concern for other parts, please let me know.

File: keras/layers/local.py [hunk 7]
```diff
@@ -392,33 +433,60 @@ def call(self, inputs):
...
+                output = K.reshape(output,
+                                   (-1, filters, self.output_row, self.output_col))
+                # shape: batch, row, col, filters
+                output = K.permute_dimensions(output, (0, 2, 3, 1))
 
         if self.use_bias:
             if self.data_format == 'channels_first':
-                output += K.reshape(self.bias,
-                                    (1, filters, self.output_row, self.output_col))
+                bias_shape = (1, filters, self.output_row, self.output_col) if K.backend() != 'cntk' else (filters, self.output_row, self.output_col)
```
Comment (TimZaman) [inline keras/layers/local.py]: Can't you solve this discrepancy in the cntk backend?

**Comment by ahundt** [comment]
I agree with @Dref360 perhaps The code causing those if statements could also be moved to the backend?

**Comment by souptc** [comment]
Thanks @Dref360 and @ahundt for the comments. we will refactor the implementation to make it looks better.

**Comment by souptc** [comment]
@Dref360 @ahundt , we have push new changes, which avoid the if statements in keras core code, by wrapper them in the backends. Could you review them? and please let me know if you have more comments.

File: keras/layers/local.py (hunk 7 above)
Comment (souptc) [inline keras/layers/local.py]: Thanks, we update the implementation by wrapper those in K.bias_add.

**Comment by ahundt** [comment]
I've only skimmed, but it is impressive that a whole new backend can be added with just a net of 2600 additional lines!

what's the reason for the new `bias_shape` param?

**Comment by souptc** [comment]
The existing bias_add only support broadcast a bias vector across axis (by extend bias shape with 1s), but for the bias in the local connection layer, bias itself is a tensor( like (output_rows, output_cols, filters) for local connect 2d). We want to take the bias_add advantage to help us deal with stuff like data_format/broadcast/..., so we passed in the shape of the bias tensor to make bias_add support it.

File: tests/keras/backend/backend_test.py [hunk 8]
```diff
@@ -16,66 +19,152 @@ def check_dtype(var, dtype):
...
-
-    assert zth.shape == ztf.shape
-    assert_allclose(zth, ztf, atol=1e-05)
+    for i in range(len(z_list) - 1):
+        assert z_list[i].shape == z_list[i + 1].shape
+        assert_allclose(z_list[i], z_list[i + 1], atol=1e-05)
 
 
 class TestBackend(object):
+    '''maybe need special handle for cntk'''
```
Comment (Dref360) [inline tests/keras/backend/backend_test.py]: Still relevant? If so, explain the diff

File: .travis.yml [hunk 9]
```diff
@@ -49,6 +53,24 @@ install:
 
   # install TensorFlow (CPU version).
   - pip install tensorflow
+  
+  # install cntk
+  - if [[ "$TRAVIS_PYTHON_VERSION" == "2.7" ]]; then
+      wget http://cntk.ai/PythonWheel/ForKeras/CPUOnly/cntk-2.0rc3-cp27-cp27mu-linux_x86_64.whl;
+      pip install cntk-2.0rc3-cp27-cp27mu-linux_x86_64.whl;
```
Comment (lukeyeager) [inline .travis.yml]: If you're not validating the checksum, why not just `pip install http://...` directly?

File: tests/keras/activations_test.py [hunk 10]
```diff
@@ -136,6 +136,10 @@ def test_elu():
     assert_allclose(result, test_values, rtol=1e-05)
 
     negative_values = np.array([[-1, -2]], dtype=K.floatx())
+    # cntk can't re-binding the input shape, so creat the model again to test with different batch size
```
Comment (lukeyeager) [inline tests/keras/activations_test.py]: typos

File: .travis.yml (hunk 9 above)
Comment (lukeyeager) [inline .travis.yml]: Also, looks like you can upgrade from `2.0rc3` to `2.0`:
https://docs.microsoft.com/en-us/cognitive-toolkit/setup-linux-python

Comment (Dref360) [inline .travis.yml]: Why is there a ForKeras section? Is there major changes between the ForKeras and the NotForKeras version?

Comment (souptc) [inline .travis.yml]: Yes, I should update this install command. The wheel under "ForKeras" section is created 2 days before CNTK 2.0 GA release, so at that time, we make this internal build public to make it work with keras travis ci pipeline. I will update it to use the official CNTK 2.0 build. Thanks for point this out.

File: tests/keras/backend/backend_test.py (hunk 8 above)
Comment (souptc) [inline tests/keras/backend/backend_test.py]: No, I should remove this one.

File: keras/layers/recurrent.py [hunk 11]
```diff
@@ -191,6 +188,9 @@ def __init__(self, return_sequences=False,
         super(Recurrent, self).__init__(**kwargs)
         self.return_sequences = return_sequences
         self.go_backwards = go_backwards
+        if K.backend() == 'cntk' and stateful:
+            raise ValueError('Stateful RNN is not support with CNTK currently.')
```
Comment (fchollet) [inline keras/layers/recurrent.py]: "not currently supported with CNTK".

File: tests/keras/backend/backend_test.py [hunk 12]
```diff
@@ -16,66 +19,152 @@ def check_dtype(var, dtype):
...
+    # Keras function return a list, take the first output
+    assert len(zc) == 1
+    zc = zc[0]
+
+    assert ztf.shape == zc.shape
+    assert_allclose(ztf, zc, atol=1e-05)
+
+
+def check_cross_entropy_with_valid_probability_distribution():
+    xval = np.asarray([[0.26157712, 0.0432167], [-0.43380741, 0.30559841], [0.20225059, -0.38956559], [-0.13805378, 0.08506755]], dtype=np.float32)
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Line too long

File: tests/keras/backend/backend_test.py [hunk 13]
```diff
@@ -16,66 +19,152 @@ def check_dtype(var, dtype):
...
+    xtf = KTF.variable(xval)
+    xth = KTH.variable(xval)
+    xc = KC.placeholder((4, 2))
+
+    yval = np.asarray([[0.46221867, 0.53778133], [0.51228984, 0.48771016], [0.64916514, 0.35083486], [0.47028078, 0.52971922]], dtype=np.float32)
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Line too long

File: tests/keras/backend/backend_test.py [hunk 14]
```diff
@@ -16,66 +19,152 @@ def check_dtype(var, dtype):
...
+    ytf = KTF.variable(yval)
+    yth = KTH.variable(yval)
+    yc = KC.placeholder((4, 2))
+
+    ztf = KTF.eval(KTF.categorical_crossentropy(xtf, ytf, from_logits=True))
+    zth = KTH.eval(KTH.categorical_crossentropy(xth, yth, from_logits=True))
+
+    func_cntk = KC.function([xc, yc], [KC.categorical_crossentropy(xc, yc, from_logits=True), ])
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Line too long

File: tests/keras/backend/backend_test.py [hunk 15]
```diff
@@ -126,29 +240,47 @@ def test_batch_dot_shape(self):
...
+        cntk_check_single_tensor_operation('resize_images', (20, 40, 40, 3),
+                                           height_factor=3,
+                                           width_factor=2,
+                                           data_format='channels_last')
+        check_single_tensor_operation('temporal_padding', (4, 3, 3), BACKENDS)
+        check_single_tensor_operation('temporal_padding', (4, 3, 3), BACKENDS, padding=(2, 2))
+        check_single_tensor_operation('spatial_2d_padding', (4, 4, 3, 3), BACKENDS)
+        check_single_tensor_operation('spatial_2d_padding', (4, 4, 3, 3), BACKENDS, padding=((2, 2), (2, 2)))
+        check_single_tensor_operation('spatial_3d_padding', (4, 4, 3, 3, 3), BACKENDS)
+        check_single_tensor_operation('spatial_3d_padding', (4, 4, 3, 3, 3), BACKENDS, padding=((2, 2), (2, 2), (2, 2)))
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Past few lines too long

File: keras/layers/advanced_activations.py [hunk 16]
```diff
@@ -202,7 +202,7 @@ def __init__(self, theta=1.0, **kwargs):
         self.theta = K.cast_to_floatx(theta)
 
     def call(self, inputs, mask=None):
-        return inputs * K.cast(inputs > self.theta, K.floatx())
+        return inputs * K.cast(K.greater(inputs, self.theta), K.floatx())
```
Comment (n17s) [inline keras/layers/advanced_activations.py]: In principle, we can override `>`, `<`, `>=`, and `<=`. The only one we cannot really override is == which cntk has reserved for C++-side object equality so we can write things like 
```python
for tensor in list_of_tensors:
...
```
This is the reason why we have not introduced infix comparison operators. The risk is someone will use == to mean cntk.equal leading to unexpected behavior, while the benefit is syntactic sugar. For now, we have opted to avoid these surprises.

File: keras/backend/tensorflow_backend.py [hunk 17]
```diff
@@ -3334,13 +3334,17 @@ def pool3d(x, pool_size, strides=(1, 1, 1), padding='valid',
     return _postprocess_conv3d_output(x, data_format)
 
 
-def bias_add(x, bias, data_format=None):
+def bias_add(x, bias, data_format=None, bias_shape=None):
```
Comment (souptc) [inline keras/backend/tensorflow_backend.py]: The purpose of this argument is to enhance the bias_add method to support add bias as a multi-dimension tensor. The existing bias_add only support to add a bias which is 1D tensor, but the bias we have in local_conv layer is a multi-dimension tensor (like (output_row, output_col, filters) for local-2d conv), so we need to know what is the target shape we expect the tensor to be before add to the output, that's why I add this argument. Otherwise we will have to put the handle of data_format and broadcast in the LocalConnected layer.

Or should we give it a better name?

File: keras/backend/tensorflow_backend.py [hunk 18]
```diff
@@ -3632,3 +3644,120 @@ def foldr(fn, elems, initializer=None, name=None):
         Same type and shape as initializer
     """
     return tf.foldr(fn, elems, initializer=initializer, name=name)
+
+
+def dropout_on_input(inputs, dim, timesteps, dropout_value, training):
```
Comment (souptc) [inline keras/backend/tensorflow_backend.py]: The original purpose of this function is because cntk have issue when repeat the dropout with time-step axis when timesteps is not a specified dimension, so we create this function to wrapper the difference. After discussion, we decide to use a workaround in cntk backend, so now we don't need this change.

File: keras/engine/topology.py [hunk 19]
```diff
@@ -1088,7 +1088,7 @@ def add_loss(self, losses, inputs=None):
         if hasattr(self, '_losses'):
             self._losses += losses
         # Update self._per_input_updates
-        if inputs == []:
+        if inputs is []:
```
Comment (souptc) [inline keras/engine/topology.py]: It is not un-related... Like Nikos explained in the comments below, CNTK reserve the "==" operator for c++ side object equality, so here "input == []" will led to un-expected result, that's why I update with "input is []".....

File: keras/layers/recurrent.py (hunk 11 above)
Comment (souptc) [inline keras/layers/recurrent.py]: thanks, fixed.

File: tests/keras/backend/backend_test.py (hunk 12 above)
Comment (souptc) [inline tests/keras/backend/backend_test.py]: fixed

File: tests/keras/backend/backend_test.py (hunk 13 above)
Comment (souptc) [inline tests/keras/backend/backend_test.py]: fixed

File: tests/keras/backend/backend_test.py (hunk 14 above)
Comment (souptc) [inline tests/keras/backend/backend_test.py]: fixed.

File: tests/keras/backend/backend_test.py (hunk 15 above)
Comment (souptc) [inline tests/keras/backend/backend_test.py]: fixed.

**Comment by souptc** [comment]
@fchollet , I just update the implementation which resolve the most comments. For the "bias_shape", I explained the purpose of the change, do you have any suggestion about a better solution?

File: keras/engine/topology.py [hunk 20]
```diff
@@ -2774,7 +2774,8 @@ def save_weights_to_hdf5_group(f, layers):
         weight_values = K.batch_get_value(symbolic_weights)
         weight_names = []
         for i, (w, val) in enumerate(zip(symbolic_weights, weight_values)):
-            if hasattr(w, 'name') and w.name:
+            # cntk does not gurantee the parameter name is unique, so prefer the second approach here.
+            if hasattr(w, 'name') and w.name and K.backend() != 'cntk':
```
Comment (souptc) [inline keras/engine/topology.py]: Maybe some misunderstanding, the original purpose is to use autogenerated names for CNTK while keep the TF and Theano behavior unchanged.....  But after discussion, we decide use the same naming strategy as Theano for cntk, to make it more compatible. Has update the implementation.

**Comment by souptc** [comment]
And for the maintainance, yes, CNTK team will definitely commit to maintain the cntk backend in the future.

File: keras/engine/topology.py (hunk 19 above)
Comment (n17s) [inline keras/engine/topology.py]: The above can be written as
```python 
if [] == inputs:
```
without triggering CNTK's operator `==`. Using `is` has its own issues.
```python
>>> a = [1]
>>> a.pop()
1
>>> a == []
True
>>> a is []
False
```

Comment (souptc) [inline keras/engine/topology.py]: fix it by add a check for list type. please let me know if you have any comments about it.

File: keras/backend/tensorflow_backend.py (hunk 17 above)
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Wouldn't the `bias` argument already have a shape, that you could just read?

File: LICENSE [hunk 21]
```diff
@@ -8,6 +8,10 @@ All contributions by Google:
 Copyright (c) 2015, Google, Inc.
 All rights reserved.
 
+All contributions by Microsoft:
+Copyright (c) 2015, Microsoft, Inc.
```
Comment (souptc) [inline LICENSE]: thanks, fixed.

**Comment by souptc** [comment]
clean the cntk_backend file with style issue.

**Comment by souptc** [comment]
fix the bias_shape issue.

File: keras/backend/cntk_backend.py [hunk 22]
```diff
@@ -0,0 +1,2016 @@
...
+
+# cntk doesn't support gradient as symbolic op, to hook up with keras model,
+# we will create gradient as a constant placeholder, here use this global
+# map to keep the mapping from grad placeholder to parameter
+grad_parameter_dict = {}
+
+NAME_SCOPE_STACK = []
+
+
+@contextmanager
```
Comment (fchollet) [inline keras/backend/cntk_backend.py]: `^M`?

File: keras/backend/tensorflow_backend.py [hunk 23]
```diff
@@ -3346,28 +3346,41 @@ def bias_add(x, bias, data_format=None):
         Output tensor.
 
     # Raises
-        ValueError: In case of invalid `data_format` argument.
+        ValueError: In case of invalid `data_format` argument, or the input bias dimension is not expect
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Can you rephrase this? The meaning is unclear. Additionally, please make sure to introduce line breaks to keep line length reasonable.

File: keras/backend/tensorflow_backend.py [hunk 24]
```diff
@@ -3346,28 +3346,41 @@ def bias_add(x, bias, data_format=None):
...
     if data_format not in {'channels_first', 'channels_last'}:
         raise ValueError('Unknown data_format ' + str(data_format))
+    bias_shape = int_shape(bias)
+    if len(bias_shape) != 1 and len(bias_shape) != ndim(x) - 1:
+        raise ValueError('Unexpected bias dimensions %d, expect to be 1 or %d dimensions'
+                         % (len(bias_shape), ndim(x)))
     if ndim(x) == 5:
         if data_format == 'channels_first':
-            x += reshape(bias, (1, int_shape(bias)[0], 1, 1, 1))
+            shape = (bias_shape[0], 1, 1, 1) if len(bias_shape) == 1 else (bias_shape[3],) + bias_shape[:3]
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Introduce an `if` block to avoid a very long line.

File: keras/backend/tensorflow_backend.py [hunk 25]
```diff
@@ -3346,28 +3346,41 @@ def bias_add(x, bias, data_format=None):
...
+            x += reshape(bias, (1,) + shape)
         elif data_format == 'channels_last':
-            x += reshape(bias, (1, 1, 1, 1, int_shape(bias)[0]))
+            shape = (1, 1, 1, bias_shape[0]) if len(bias_shape) == 1 else bias_shape
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Same.

File: keras/backend/tensorflow_backend.py [hunk 26]
```diff
@@ -3346,28 +3346,41 @@ def bias_add(x, bias, data_format=None):
...
+            x += reshape(bias, (1,) + shape)
     elif ndim(x) == 4:
         if data_format == 'channels_first':
-            x += reshape(bias, (1, int_shape(bias)[0], 1, 1))
+            shape = (bias_shape[0], 1, 1) if len(bias_shape) == 1 else (bias_shape[2],) + bias_shape[:2]
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Same.

File: keras/backend/tensorflow_backend.py [hunk 27]
```diff
@@ -3346,28 +3346,41 @@ def bias_add(x, bias, data_format=None):
...
-                               data_format='NHWC')
+            if len(bias_shape) == 1:
+                x = tf.nn.bias_add(x, bias,
+                                   data_format='NHWC')
+            else:
+                x += reshape(bias, (1,) + bias_shape)
     elif ndim(x) == 3:
         if data_format == 'channels_first':
-            x += reshape(bias, (1, int_shape(bias)[0], 1))
+            shape = (int_shape(bias)[0], 1) if len(bias_shape) == 1 else (bias_shape[1],) + bias_shape[:1]
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Same.

File: keras/backend/tensorflow_backend.py [hunk 28]
```diff
@@ -3346,28 +3346,41 @@ def bias_add(x, bias, data_format=None):
...
+            x += reshape(bias, (1,) + shape)
         elif data_format == 'channels_last':
-            x += reshape(bias, (1, 1, int_shape(bias)[0]))
+            shape = (1, int_shape(bias)[0]) if len(bias_shape) == 1 else bias_shape
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Same.

File: keras/backend/tensorflow_backend.py [hunk 29]
```diff
@@ -3632,3 +3645,100 @@ def foldr(fn, elems, initializer=None, name=None):
...
+        kernel: the unshared weight for convolution, with shape (output_length, feature_dim, filters)
+        kernel_size: a tuple of a single integer, specifying the length of the 1D convolution window
+        strides: a tuple of a single integer, specifying the stride length of the convolution
+        data_format: the data format, channels_first or channels_last
+
+    # Returns
+        the tensor after 1d conv with un-shared weights, with shape (batch_size, output_lenght, filters)
+
+    # Raises
+        ValueError: if `data_format` is neither `channels_last` or `channels_first`.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please introduce line breaks in the docstring to avoid very long lines.

File: keras/backend/tensorflow_backend.py [hunk 30]
```diff
@@ -3632,3 +3645,100 @@ def foldr(fn, elems, initializer=None, name=None):
...
+        data_format: the data format, channels_first or channels_last
+
+    # Returns
+        A 4d tensor with shape: (batch_size, filters, new_rows, new_cols)
+        if data_format='channels_first'
+        or 4D tensor with shape: (batch_size, new_rows, new_cols, filters)
+        if data_format='channels_last'.
+
+    # Raises
+        ValueError: if `data_format` is neither `channels_last` or `channels_first`.
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Please introduce line breaks in the docstring to avoid very long lines.

File: keras/backend/theano_backend.py [hunk 31]
```diff
@@ -2050,21 +2050,31 @@ def bias_add(x, bias, data_format=None):
         data_format = image_data_format()
     if data_format not in {'channels_first', 'channels_last'}:
         raise ValueError('Unknown data_format ' + str(data_format))
+    if ndim(bias) != 1 and ndim(bias) != ndim(x) - 1:
+        raise ValueError('Unexpected bias dimensions %d, expect to be 1 or %d dimensions'
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Line too long

File: keras/backend/theano_backend.py [hunk 32]
```diff
@@ -2050,21 +2050,31 @@ def bias_add(x, bias, data_format=None):
...
+                         % (ndim(bias), ndim(x) - 1))
+    bias_shape = tuple(bias.shape)
     if ndim(x) == 5:
         if data_format == 'channels_first':
-            x += reshape(bias, (1, bias.shape[0], 1, 1, 1))
+            shape = (bias_shape[0], 1, 1, 1) if ndim(bias) == 1 else (bias_shape[3],) + bias_shape[:3]
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Line too long

File: keras/backend/theano_backend.py [hunk 33]
```diff
@@ -2050,21 +2050,31 @@ def bias_add(x, bias, data_format=None):
...
+            x += reshape(bias, (1,) + shape)
         elif data_format == 'channels_last':
-            x += reshape(bias, (1, 1, 1, 1, bias.shape[0]))
+            shape = (1, 1, 1, bias_shape[0]) if ndim(bias) == 1 else bias_shape
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Line too long

File: keras/backend/theano_backend.py [hunk 34]
```diff
@@ -2050,21 +2050,31 @@ def bias_add(x, bias, data_format=None):
...
+            x += reshape(bias, (1,) + shape)
     elif ndim(x) == 4:
         if data_format == 'channels_first':
-            x += reshape(bias, (1, bias.shape[0], 1, 1))
+            shape = (bias_shape[0], 1, 1) if ndim(bias) == 1 else (bias_shape[2],) + bias_shape[:2]
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Line too long

File: keras/backend/theano_backend.py [hunk 35]
```diff
@@ -2050,21 +2050,31 @@ def bias_add(x, bias, data_format=None):
...
+            x += reshape(bias, (1,) + shape)
         elif data_format == 'channels_last':
-            x += reshape(bias, (1, 1, 1, bias.shape[0]))
+            shape = (1, 1, bias_shape[0]) if ndim(bias) == 1 else bias_shape
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Line too long

File: keras/backend/theano_backend.py [hunk 36]
```diff
@@ -2050,21 +2050,31 @@ def bias_add(x, bias, data_format=None):
...
+            x += reshape(bias, (1,) + shape)
     elif ndim(x) == 3:
         if data_format == 'channels_first':
-            x += reshape(bias, (1, bias.shape[0], 1))
+            shape = (bias_shape[0], 1) if ndim(bias) == 1 else (bias_shape[1],) + bias_shape[:1]
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Line too long

File: keras/backend/theano_backend.py [hunk 37]
```diff
@@ -2050,21 +2050,31 @@ def bias_add(x, bias, data_format=None):
...
+            x += reshape(bias, (1,) + shape)
         elif data_format == 'channels_last':
-            x += reshape(bias, (1, 1, bias.shape[0]))
+            shape = (1, bias_shape[0]) if ndim(bias) == 1 else bias_shape
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Line too long

File: keras/layers/local.py [hunk 38]
```diff
@@ -363,62 +352,14 @@ def compute_output_shape(self, input_shape):
...
-                    slice_col = slice(j * stride_col,
-                                      j * stride_col + self.kernel_size[1])
-                    xs.append(K.reshape(inputs[:, slice_row, slice_col, :],
-                                        (1, -1, feature_dim)))
-            x_aggregate = K.concatenate(xs, axis=0)
-            output = K.batch_dot(x_aggregate, self.kernel)
-            output = K.reshape(output,
-                               (self.output_row, self.output_col, -1, filters))
-            output = K.permute_dimensions(output, (2, 0, 1, 3))
+        output = K.local_conv2d(inputs, self.kernel, self.kernel_size, self.strides, (self.output_row, self.output_col), self.data_format)
```
Comment (fchollet) [inline keras/layers/local.py]: Line too long

**Comment by fchollet** [comment]
The failing test is a flake.

**Comment by fchollet** [comment]
I have applied a few style fixes to the backend source and merged. Thanks a lot! 👍 

While going through the backend code, I have noticed that a lot of the error messages raised were not as helpful and as clear as they should be. I suggest you go over the error messages and improve them. Typically you want to tell the user: 1) what they did (e.g. print the arguments they passed), 2) why that was wrong (e.g. something not supported), and 3) what they should do instead. Currently you are mostly doing 2) only.

This will greatly improve the user experience for CNTK Keras users, and improve the usability / ease of debugging of Keras models on CNTK.

**Comment by souptc** [comment]
Thanks François! Sure, I will go though the message tomorrow and improve them.

**Comment by ebarsoumMS** [comment]
Thanks, everybody...

**Comment by fchollet** [comment]
Huge thanks to everyone who contributed to this project! It's a big milestone.
